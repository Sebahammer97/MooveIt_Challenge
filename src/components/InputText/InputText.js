import React, { Component } from "react";
import cx from "classnames";
import PropTypes from "prop-types";

// Components
import Button from "../Button/Button";

class Input extends Component {
  // Definition of props types to this component
  static propTypes = {
    onChange: PropTypes.func,
    className: PropTypes.string,
    placeholder: PropTypes.string,
    value: PropTypes.string,
    variant: PropTypes.string,
    size: PropTypes.string,
    disabledClassName: PropTypes.string,
    disabled: PropTypes.bool,
    type: PropTypes.string.isRequired,
    minLength: PropTypes.number,
    maxLength: PropTypes.number,
    textarea: PropTypes.bool,
    reset: PropTypes.bool,
  };

  // Assigning default values for the props
  static defaultProps = {
    className: "input",
    placeholder: "",
    value: "",
    variant: "default",
    disabled: false,
    disabledClassName: "",
    size: "medium",
    type: "text",
    minLength: 0,
    maxLength: 9999,
    textarea: false,
    reset: false,
  };

  handleInputChange = (event) => {
    const { onChange, disabled } = this.props;

    if (disabled) {
      return;
    }

    if (onChange) {
      onChange(event.target.value);
    }
  };

  resetInputValue = () => {
    const { onChange } = this.props;
    if (onChange) {
      onChange("");
    }
  };

  renderChildren = () => {
    const { placeholder, children } = this.props;

    if (placeholder) {
      return placeholder;
    }

    if (children) {
      return children;
    }

    return "";
  };

  render() {
    const {
      className,
      type,
      value,
      placeholder,
      size,
      minLength,
      maxLength,
      variant,
      disabled,
      disabledClassName,
      textarea,
      reset,
    } = this.props;

    const styles = require("./style.module.css");

    const _className = cx(className, styles[size], styles[variant], {
      [styles.disabled]: disabled,
      [styles.error]:
        value.length < minLength || (maxLength && value.length > maxLength),
      [disabledClassName]: disabled,
    });

    const InputElement = textarea ? "textarea" : "input";

    return (
      <div>
        <InputElement
          className={_className}
          type={type}
          value={value}
          placeholder={placeholder}
          minLength={minLength}
          maxLength={maxLength}
          onChange={this.handleInputChange}
          disabled={disabled}
          data-testid="input-text"
        />
        {reset && value.length > 0 ? (
          <Button label="X" variant="reset" onClick={this.resetInputValue} />
        ) : null}
      </div>
    );
  }
}

export default Input;
