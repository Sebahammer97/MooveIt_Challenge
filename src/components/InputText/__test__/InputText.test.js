import React, { useState } from "react";
import ReactDOM from "react-dom";
import InputText from "../InputText";

import { render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

afterEach(cleanup);

it("input text renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<InputText type={"text"} />, div);
});

it("textarea renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<InputText type={"text"} textarea={true} />, div);
});

it("password renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<InputText type={"password"} />, div);
});

function InputTextTest({ type, textarea, variant, disabled, size }) {
  const [input, setInput] = useState("");

  return (
    <InputText
      type={type}
      textarea={textarea}
      value={input}
      onChange={(value) => setInput(value)}
      variant={variant}
      disabled={disabled}
      size={size}
    />
  );
}

it("renders input text correctly", () => {
  const utils = render(<InputTextTest type="text" />);
  const input = utils.getByTestId("input-text");
  fireEvent.change(input, { target: { value: "abcdefg" } });
  expect(input.value).toBe("abcdefg");
  fireEvent.change(input, { target: { value: "123456" } });
  expect(input.value).toBe("123456");
  fireEvent.change(input, { target: { value: "a1b2c3d4f5" } });
  expect(input.value).toBe("a1b2c3d4f5");
});

it("renders textarea correctly", () => {
  const utils = render(<InputTextTest type="text" textarea={true} />);
  const input = utils.getByTestId("input-text");
  fireEvent.change(input, { target: { value: "abcdefg" } });
  expect(input.value).toBe("abcdefg");
  fireEvent.change(input, { target: { value: "123456" } });
  expect(input.value).toBe("123456");
  fireEvent.change(input, { target: { value: "a1b2c3d4f5" } });
  expect(input.value).toBe("a1b2c3d4f5");
});

it("renders password correctly", () => {
  const utils = render(<InputTextTest type="password" />);
  const input = utils.getByTestId("input-text");
  fireEvent.change(input, { target: { value: "abcdefg" } });
  expect(input.value).toBe("abcdefg");
  fireEvent.change(input, { target: { value: "123456" } });
  expect(input.value).toBe("123456");
  fireEvent.change(input, { target: { value: "a1b2c3d4f5" } });
  expect(input.value).toBe("a1b2c3d4f5");
});

it("disable input text correctly", () => {
  const utils = render(<InputTextTest type="text" disabled={true} />);
  const input = utils.getByTestId("input-text");
  fireEvent.change(input, { target: { value: "abcdefg" } });
  expect(input).toBeDisabled();
});

it("matches snapshots input text", () => {
  const tree = renderer.create(<InputText type="text" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots textarea", () => {
  const tree = renderer
    .create(<InputText type="text" textarea={true} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots password", () => {
  const tree = renderer.create(<InputText type="password" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots disabled", () => {
  const tree = renderer
    .create(<InputText type="text" disabled={true} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots with reset", () => {
  const tree = renderer.create(<InputText type="text" reset={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots size", () => {
  const tree = renderer
    .create(
      <div>
        <InputText type="text" size="small" />{" "}
        <InputText type="text" size="medium" />{" "}
        <InputText type="text" size="large" />
      </div>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots variant styles", () => {
  const tree = renderer
    .create(
      <div>
        <InputText type="text" variant="dark" />
      </div>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
