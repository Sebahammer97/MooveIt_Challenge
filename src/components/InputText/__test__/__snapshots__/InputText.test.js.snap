// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots disabled 1`] = `
<div>
  <input
    className="input medium default disabled "
    data-testid="input-text"
    disabled={true}
    maxLength={9999}
    minLength={0}
    onChange={[Function]}
    placeholder=""
    type="text"
    value=""
  />
</div>
`;

exports[`matches snapshots input text 1`] = `
<div>
  <input
    className="input medium default"
    data-testid="input-text"
    disabled={false}
    maxLength={9999}
    minLength={0}
    onChange={[Function]}
    placeholder=""
    type="text"
    value=""
  />
</div>
`;

exports[`matches snapshots password 1`] = `
<div>
  <input
    className="input medium default"
    data-testid="input-text"
    disabled={false}
    maxLength={9999}
    minLength={0}
    onChange={[Function]}
    placeholder=""
    type="password"
    value=""
  />
</div>
`;

exports[`matches snapshots size 1`] = `
<div>
  <div>
    <input
      className="input small default"
      data-testid="input-text"
      disabled={false}
      maxLength={9999}
      minLength={0}
      onChange={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
   
  <div>
    <input
      className="input medium default"
      data-testid="input-text"
      disabled={false}
      maxLength={9999}
      minLength={0}
      onChange={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
   
  <div>
    <input
      className="input large default"
      data-testid="input-text"
      disabled={false}
      maxLength={9999}
      minLength={0}
      onChange={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`matches snapshots textarea 1`] = `
<div>
  <textarea
    className="input medium default"
    data-testid="input-text"
    disabled={false}
    maxLength={9999}
    minLength={0}
    onChange={[Function]}
    placeholder=""
    type="text"
    value=""
  />
</div>
`;

exports[`matches snapshots variant styles 1`] = `
<div>
  <div>
    <input
      className="input medium dark"
      data-testid="input-text"
      disabled={false}
      maxLength={9999}
      minLength={0}
      onChange={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`matches snapshots with reset 1`] = `
<div>
  <input
    className="input medium default"
    data-testid="input-text"
    disabled={false}
    maxLength={9999}
    minLength={0}
    onChange={[Function]}
    placeholder=""
    type="text"
    value=""
  />
</div>
`;
